app/dashboard/src/layouts/AssetsTable.tsx

```diff
@@ -412,42 +322,249 @@ export default function AssetsTable(props: AssetsTableProps) {
   const [pasteData, setPasteData] = React.useState<pasteDataModule.PasteData<
     ReadonlySet<backendModule.AssetId>
   > | null>(null)
-  const [, setQueuedAssetEvents] = React.useState<assetEvent.AssetEvent[]>([])
   const nameOfProjectToImmediatelyOpenRef = React.useRef(initialProjectName)
   const rootDirectoryId = React.useMemo(
     () => backend.rootDirectoryId(user) ?? backendModule.DirectoryId(''),
     [backend, user],
   )
-  const [assetTree, setAssetTree] = React.useState<assetTreeNode.AnyAssetTreeNode>(() => {
-    const rootParentDirectoryId = backendModule.DirectoryId('')
-    return AssetTreeNode.fromAsset(
-      backendModule.createRootDirectoryAsset(rootDirectoryId),
+
+  const rootParentDirectoryId = backendModule.DirectoryId('')
+  const rootDirectory = React.useMemo(
+    () => backendModule.createRootDirectoryAsset(rootDirectoryId),
+    [rootDirectoryId],
+  )
+
+  const enableAssetsTableBackgroundRefresh = useFeatureFlag('enableAssetsTableBackgroundRefresh')
+  const assetsTableBackgroundRefreshInterval = useFeatureFlag(
+    'assetsTableBackgroundRefreshInterval',
+  )
+  /**
+   * The expanded directories in the asset tree.
+   */
+  const [expandedDirectoryIds, setExpandedDirectoryIds] = React.useState<
+    backendModule.DirectoryId[]
+  >(() => [rootDirectory.id])
+
+  const expandedDirectoryIdsSet = React.useMemo(
+    () => new Set(expandedDirectoryIds),
+    [expandedDirectoryIds],
+  )
+
+  const createProjectMutation = useMutation(
+    backendMutationOptions(backend, 'createProject', {
+      meta: { invalidates: [['listDirectory', backend.type]] },
```

Comment:
consider extracting `['listDirectory', backend.type]` to a constant and using it here and in `directories`

(also why is it this and not `[backend.type, 'listDirectory']`? IMO i'm not sure whether it makes sense to want to invalidate all queries for a specific *method* for *both* backends at once)

Reply:
Well, it's discussible, and, as of now, I'm not opinionated on the order in `queryHash`. If we find `[backend.type, 'listDirectory']` more consistent, then I'm ok with that

Reply:
yeah. to be fair i'm not too opinionated either, but most of the other methods (especially the ones in `backendHooks.ts`) *currently* have `backend.type` first.

although i imagine it would be easy enough to switch to `backend.type` second - but i guess it's less change required (and therefore safer) to change just these instances rather than going through the entire codebase and risk missing some

Reply:
@somebody1234 here's an interesting article on Query keys: https://tkdodo.eu/blog/leveraging-the-query-function-context#object-query-keys